### PR TITLE
[vscode] support CommentThread state property

### PR DESCRIPTION
**Pull request by rschnekenbu**
#### What it does

_Add Comment#state optional property_
This adds the state optional property from Comment to improve vscode api coverage. It also displays the state in the Comment Thread widget

Fixes #12231 

Contributed on behalf of ST Microelectronics

#### How to test 
Note: there are currently some issues with menus, since theia 1.28.0. This is due to #11290, integrated in 1.28.0. Prior to this change, the comment sample works on theia 1.27.0. So to test this change, I cherry-picked my commit on top of 1.27.0 tag. 
This is the same case as for PR https://github.com/eclipse-theia/theia/pull/12007.
1. Install forked extension [#comment-sample-0.0.1](https://github.com/rschnekenbu/vscode-extension-samples/releases/download/comment-with-state/forked-comment-api-example-0.0.1.vsix) vsix file - [Source](https://github.com/rschnekenbu/vscode-extension-samples/tree/rschnekenbu/comment-state/comment-sample)
2. Create a comment using the side bar while comparing 2 test files and reply with some comment with toggle or reply. 

3. The resolved state should be displayed along the thread label.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- [x] As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/main/browser/comments/comments-main.ts [hunk 1]
```diff
@@ -124,6 +124,19 @@ export class CommentThreadImpl implements CommentThread, Disposable {
...
     readonly onDidChangeCollapsibleState = this.onDidChangeCollapsibleStateEmitter.event;
 
+    private _state: CommentThreadState | undefined;
+    get state(): CommentThreadState | undefined {
+        return this._state;
+    }
+
+    set state(newState: CommentThreadState | undefined) {
+        this._state = newState;
+        this.onDidChangeStateEmitter.fire(this._state);
```
Comment (tsmaeder) [inline packages/plugin-ext/src/main/browser/comments/comments-main.ts:134]: Does it make sense to check whether the state has really changed before notifying?

File: packages/plugin-ext/src/main/browser/comments/comment-thread-widget.tsx [hunk 2]
```diff
@@ -231,7 +235,8 @@ export class CommentThreadWidget extends BaseWidget {
             if (this._commentThread.comments && this._commentThread.comments.length) {
                 const onlyUnique = (value: Comment, index: number, self: Comment[]) => self.indexOf(value) === index;
                 const participantsList = this._commentThread.comments.filter(onlyUnique).map(comment => `@${comment.userName}`).join(', ');
-                label = `Participants: ${participantsList}`;
+                const resolved = (this._commentThread.state === CommentThreadState.Resolved) ? '(Resolved)' : '(Unresolved)';
```
Comment (tsmaeder) [inline packages/plugin-ext/src/main/browser/comments/comment-thread-widget.tsx:238]: I was confused by the variable name `resolved`. Why not call it `state` or `resolutionState`?

File: packages/plugin-ext/src/plugin/comments.ts [hunk 3]
```diff
@@ -276,6 +278,18 @@ export class ExtHostCommentThread implements theia.CommentThread, theia.Disposab
         this._onDidUpdateCommentThread.fire();
     }
 
+    private threadState?: theia.CommentThreadState;
```
Comment (tsmaeder) [inline packages/plugin-ext/src/plugin/comments.ts:281]: The pattern used in this file is to name the private var `_state`.

File: packages/plugin-ext/src/plugin/comments.ts [hunk 4]
```diff
@@ -276,6 +278,18 @@ export class ExtHostCommentThread implements theia.CommentThread, theia.Disposab
...
+
+    get state(): theia.CommentThreadState {
+        return this.threadState!;
+    }
+
+    set state(newState: theia.CommentThreadState) {
+        this.threadState = newState;
```
Comment (tsmaeder) [inline packages/plugin-ext/src/plugin/comments.ts:288]: Should we check whether the value is really modified here?

**Comment by tsmaeder** [comment]
@rschnekenbu do we have an issue for the problems with the menus you mentioned? If not, could you open one, please?

File: packages/plugin-ext/src/main/browser/comments/comments-main.ts (hunk 1 above)
Comment (rschnekenbu) [reply packages/plugin-ext/src/main/browser/comments/comments-main.ts]: Good catch! Resolved with 830e637

File: packages/plugin-ext/src/main/browser/comments/comment-thread-widget.tsx (hunk 2 above)
Comment (rschnekenbu) [reply packages/plugin-ext/src/main/browser/comments/comment-thread-widget.tsx]: Solved with 830e637e, now renamed resolutionState.

File: packages/plugin-ext/src/plugin/comments.ts (hunk 3 above)
Comment (rschnekenbu) [reply packages/plugin-ext/src/plugin/comments.ts]: Thanks! renamed with 830e637e

File: packages/plugin-ext/src/plugin/comments.ts (hunk 4 above)
Comment (rschnekenbu) [reply packages/plugin-ext/src/plugin/comments.ts]: This is now checked with 830e637e to avoid unneeded notifications.

**Comment by rschnekenbu** [comment]
> @rschnekenbu do we have an issue for the problems with the menus you mentioned? If not, could you open one, please?

The issue https://github.com/eclipse-theia/theia/issues/12452 has been created to track those problems about menus.

**Review by AlexandraBuzila — APPROVED** [review]
LGTM and works as expected (tested on top of v1.27)

Thanks Remi!

**Comment by rschnekenbu** [comment]
Closed in favor of eclipse-theia PR.

**Pull request closed without merging**